exatomic/adf/nmr/output.py

```diff
@@ -0,0 +1,135 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2015-2018, Exa Analytics Development Team
+# Distributed under the terms of the Apache License 2.0
+"""
+ADF NMR Output Editor
+#############################
+Editor class for parsing the NMR data from an ADF calculation
+"""
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import division
+from collections import defaultdict
+import re
+import six
+import numpy as np
+import pandas as pd
+from io import StringIO
+from exa.util.units import Length
+from exa import TypedMeta
+from exatomic.base import sym2z
+#from exatomic.algorithms.basis import lmap, enum_cartesian
+#from exatomic.algorithms.numerical import dfac21
+from exatomic.core.atom import Atom
+from exatomic.core.tensor import NMRShielding, JCoupling
+#from exatomic.core.basis import BasisSet, BasisSetOrder
+#from ..core.orbital import Orbital, Excitation, MOMatrix
+from exatomic.adf.editor import Editor
+
+class OutMeta(TypedMeta):
+    atom = Atom
+    nmr_shielding = NMRShielding
+    j_coupling = JCoupling
+
+class Output(six.with_metaclass(OutMeta, Editor)):
+    """ADF NMR parser"""
+    def parse_atom(self):
+        # use the regex instead of find because we have a similar search string in an nmr and
+        # cpl calculation for the nuclear coordinates
+        _reatom = "(?i)NUCLEAR COORDINATES"
+        found = self.regex(_reatom, keys_only=True)
+        #if len(found) > 1:
+        #    raise NotImplementedError("We can only parse outputs from a single NMR calculation")
+        atom = []
+        for idx, val in enumerate(found):
+            start = val + 3
+            stop = start
+            while self[stop].strip(): stop += 1
+            # a bit of a hack to make sure that there is no formatting change depending on the
+            # number of atoms in the molecule as the index is right justified so if there are
+            # more than 100 atoms it will fill the alloted space for the atom index and change the
+            # delimitter and therefore the number of columns
+            self[start:stop] = map(lambda x: x.replace('(', ''), self[start:stop])
+            df = self.pandas_dataframe(start, stop, ncol=5)
+            df.columns = ['symbol', 'set', 'x', 'y', 'z']
+            for c in ['x', 'y', 'z']: df[c] *= Length['Angstrom', 'au']
+            df['Z'] = df['symbol'].map(sym2z)
+            df['frame'] = idx
+            # remove the trailing chracters from the index
+            df['set'] = list(map(lambda x: x.replace('):', ''), df['set']))
+            df['set'] = df['set'].astype(int) - 1
+            atom.append(df)
+        self.atom = pd.concat(atom, ignore_index=True)
+
+    def parse_nmr_shielding(self):
+        _reatom = "N U C L E U S :"
```

Comment:
I guess we could, but I don't see much of a gain in doing this.

Reply:
Exposing those constants at the class level allows the 'poweruser' to modify them. Could be useful for testing and debugging. Exposing those at the class level instantiates those values once instead of every time those methods are called. There are still other useful things about this related to class inheritance.